✨ feat : 온보딩 적성 AI 추천 기능 전체 구현

sooyeoneo opened the pull request:
## ☀️ 작업 목적

* Spring AI(GPT-4o) 기반 **AI 적성 추천** 및 **사용자 수동 적성 저장** 기능을 백엔드에 구현했습니다.
* 온보딩 단계에서 청년 사용자의 진로 탐색을 돕는 핵심 적성 추천 로직 및 관련 인프라 구조 설계를 완료하였습니다.

---

## 📄 주요 작업 내역

#### 1. 온보딩 **적성 AI 추천** 및 **수동 적성 저장** 기능 개발

#### 2. Spring AI 기반 AI 적성 검사 프로세스 및 **Entity/DTO/Repository/Service/Controller** 구현




* **Entity/Enum**

  * `UserAptitude` 엔티티: 회원별 적성 결과 및 테스트 횟수(최대 5회) 관리, BaseEntity 상속
  * `AptitudeTestHistory` 엔티티: AI 적성 검사 이력(질문-답변-AI응답) 저장
  * `AptitudeType` Enum: 5가지 적성 타입(자연친화, 역사문화, 예술창작, 커뮤니티, 디지털기술) 정의 및 설명 포함

* **Repository**

  * `UserAptitudeRepository`: 회원별 적성 정보 조회, 존재여부 확인
  * `AptitudeTestHistoryRepository`: 회원별 검사 이력 조회/전체 삭제 지원

* **DTO**

  * `AptitudeAnswerRequestDto`: 답변 제출 요청
  * `AptitudeQuestionResponseDto`: AI 질문 응답
  * `AptitudeTestStartResponseDto`: 검사 시작 응답
  * `AptitudeTextProgressResponseDto`: 검사 진행 상태 응답
  * `AptitudeTestResultResponseDto`: 최종 결과 응답
  * `CanRetakeTestResponseDto`: 재검사 가능 여부 응답
  * → 모든 ResponseDto는 불변(생성자) 패턴 적용

* **Service**

  * `AptitudeAiService`: 5개 질문 Pool 관리, AI 프롬프트 기반 답변 분석, 점수 계산 및 최종 적성 결정, BaseAiService 상속(재시도/예외처리)
  * `AptitudeService`: 적성 검사 시작/답변 제출/결과 저장, Redis 기반 진행 상태 관리(24시간 TTL), 5회 제한, 수동 적성 선택, CustomException 처리
  * 메서드별 @Transactional 적용

* **Controller**

  * 주요 엔드포인트:

    * `POST /api/ai/aptitude/test/start`: 검사 시작
    * `POST /api/ai/aptitude/test/answer`: 답변 제출
    * `POST /api/ai/aptitude/select`: 수동 적성 선택
    * `GET /api/ai/aptitude/can-retake`: 재검사 가능 여부
  * `@PreAuthorize("hasRole('USER')")` 일반회원 권한 적용
  * `CommonResponseDto<T>` 응답 컨벤션 적용

* **Redis 설정**

  * 범용 `aiAptitudeRedisTemplate` 빈 추가 및 `@Qualifier` 지정

* **테스트 및 프롬프트**

  * `ai-aptitude.http` HTTP API 테스트 코드 추가(Authorization Bearer)
  * `aptitude-test.txt` AI 적성 분석 프롬프트 추가

---

## 🛠️ 추가 수정 작업

* ~~질문 Pool 관리 방식 개선: 코드 하드코딩 → 추후 외부 파일/DB로 유연화 예정~~
  * 질문 생성 로직 개선: 정적 질문 Pool → 대화 맥락 기반 AI 동적 질문 생성 (2025.09.15.)
* Redis 활용: AI 적성 테스트 단계 및 세션 TTL(24시간) 자동 만료 적용
* 테스트 재시도 횟수 체크 및 남은 횟수 반환 API 추가
* 적성 Enum(`AptitudeType`) 구조화 및 한글명/설명 포함

---

## 📌 참고 및 TODO

* 커밋별 상세 구현/구조는 각 주요 항목에 반영됨.
* 이후 추가 반영 시 TODO 체크리스트 기반으로 순차 개선 예정.

* **예외처리(GlobalExceptionHandler)**
  : 현재 `CustomException` 및 `ErrorCode`로 분기 처리만 되어 있음.
  → *ControllerAdvice 기반 전역 예외처리 및 일관된 에러 메시지 반환 필요* `2025.07.21T23:30`
  : 코멘트 반영하여 `@PreAuthorize("hasRole('USER')")`삭제.
  → *로컬 크리에이터로 로그인 시 적성 검사 가능한 상태를 불가능한 상태로 예외 처리* `2025.07.22T02:00`
* **입력값 유효성 검증**
  : Controller 단에서 `@Valid`, `@NotBlank`, `@NotNull` 등 미적용
  → *입력 DTO에 어노테이션 적용 및 ValidationError 처리 필수* `2025.07.31T07:08`
* **Swagger(OpenAPI) 문서화**
  : `@Operation`, `@ApiResponses`, `@Parameter` 등 API 명세 미정의
  → *팀/외부 협업을 위한 API 문서화 빠른 적용 필요* `2025.07.31T07:08`
* **AI 응답 파싱 고도화**
  : 현재 String 기반 키워드 매칭
  → *프롬프트에 JSON 포맷 명시 + ObjectMapper 등으로 파싱 구조 개선 필요 (정확도/유지보수성↑)* `2025.09.03T11:51`
* **수동 선택/AI 테스트 횟수 분리**
  : `UserAptitude.testCount`가 수동 적성 저장 시에도 증가
  → *AI 테스트만 카운트하도록 엔티티/메서드 분리 리팩토링 필요*
* **AptitudeTestHistory 활용성 확장**
  : AI 응답 전체 저장만 가능, 추후 적성 추천 근거/설명 등 결과 DTO에 추가 필요
* **사용자 UX**
  : AI 추천 후, 프로그램 예시 제공/최종 확정 버튼(완료 및 시작하기) 분리 로직 미구현
  → *결과 임시 반환 후, 사용자 동의 후 최종 저장 구조로 개선 필요*
* **프로그램 추천**
  : 적성별 체험 프로그램 5\~7개 예시 제공 API 및 DTO 확장 필요
* **HashMap 최적화**
  : 적성 검사 AI 응답 분석에 정규식 패턴 매칭 로직 추가로 향후 HashMap 최적화 예정 `2025.07.22T3:30`

---

### ⚙️ 기타

* 트랜잭션 격리 수준(기본 READ\_COMMITTED, 동시성 문제/필요시 추가 지정 고려)
* 민감정보(.env, .gitignore) 및 운영환경 변수 적용 완료
* HashMap 최적화

---

**리뷰 및 피드백 환영합니다! 🤗**

sooyeoneo committed:
✨ feat : 적성 검사 Entity 및 Enum 구현

- UserAptitude Entity 구현
  - 회원별 적성 정보 저장 (1:1 관계)
  - 테스트 횟수 추적 (최대 5회 제한)
  - BaseEntity 상속으로 생성/수정 시간 자동 관리

- AptitudeTestHistory Entity 구현
  - 적성 검사 이력 저장 (N:1 관계)
  - 질문-답변-AI응답 기록

- AptitudeType Enum 구현
  - 5가지 적성 타입 정의 (자연친화, 역사문화, 예술창작, 커뮤니티, 디지털기술)
  - 각 타입별 제목과 설명 포함

#65

sooyeoneo committed:
✨ feat : 적성 검사 Repository 구현

- UserAptitudeRepository
  - 회원별 적성 조회 (findByMember)
  - 적성 존재 여부 확인 (existsByMember)

- AptitudeTestHistoryRepository
  - 회원별 검사 이력 조회 (오름차순 정렬)
  - 회원별 이력 전체 삭제

#65

sooyeoneo committed:
✨ feat : 적성 검사 DTO 구현

- RequestDto (요청)
  - AptitudeAnswerRequestDto: 사용자 답변 제출

- ResponseDto (응답)
  - AptitudeQuestionResponseDto: AI 질문 정보를 사용자에게 답변 응답
  - AptitudeTestStartResponseDto: 검사 시작 응답
  - AptitudeTextProgressResponseDto: 진행 상태 응답
  - AptitudeTestResultResponseDto: 최종 결과 응답
  - CanRetakeTestResponseDto: 재검사 가능 여부 응답

* 모든 ResponseDto는 불변 객체로 구현 (생성자 사용)

#64 #65

sooyeoneo committed:
✨ feat : 적성 검사 Service 구현

- AptitudeAiService
  - 5개 질문 Pool 정적 관리
  - AI 프롬프트 기반 답변 분석
  - 적성 타입별 점수 계산 및 최종 적성 결정
  - BaseAiService 상속으로 재시도 및 예외처리

- AptitudeService
  - 검사 시작/답변 제출/결과 저장 플로우
  - Redis 기반 검사 진행 상태 관리 (24시간 TTL)
  - 최대 5회 재검사 제한 로직
  - 수동 적성 선택 기능
  - CustomException 기반 예외처리

* @Transactional 메서드별 적용

#64 #65 #66

sooyeoneo committed:
✨ feat : 적성 검사 Controller 구현

- API 엔드포인트
  - POST /api/ai/aptitude/test/start: 검사 시작
  - POST /api/ai/aptitude/test/answer: 답변 제출
  - POST /api/ai/aptitude/select: 수동 적성 선택
  - GET /api/ai/aptitude/can-retake: 재검사 가능 여부

- Spring Security 적용
  - @PreAuthorize(\"hasRole('USER')\") 일반회원 권한 체크
  - MemberDetails 통한 회원 정보 획득

- CommonResponseDto 표준 응답 형식 적용

#63 #66

sooyeoneo committed:
🔧 add : Redis 범용 aiAptitudeRedisTemplate 빈 추가

- RedisTemplate<String, String> 빈 추가
- @Qualifier("aiAptitudeRedisTemplate") 빈에 추가 구분자를 붙여 명시적으로 지정
- 적성 검사 진행 상태 관리용
- 기존 blacklist, emailVerified 템플릿과 별도 관리

#64

sooyeoneo committed:
✅ test : 적성 검사 HTTP 테스트 및 AI 프롬프트 추가

- ai-aptitude.http
  - 적성 검사 시작/답변/선택/재검사 API 테스트
  - Authorization Bearer 토큰 기반

- aptitude-test.txt
  - 적성 분석 AI 프롬프트 추가

#67

sooyeoneo committed:
💡 comment : AiTestController 주석 처리 (추후 삭제 예정)

sooyeoneo committed:
💡 comment : 필요 없는 주석 삭제

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java:
```diff
@@ -0,0 +1,204 @@
+		String redisKey = REDIS_KEY_PREFIX + memberId;
+		String currentStepStr = aiAptitudeRedisTemplate.opsForValue().get(redisKey);
+		if (currentStepStr == null) {
+			throw new CustomException(ErrorCode.INVALID_INPUT_VALUE, "진행 중인 테스트가 없습니다.");
```
별거는 아니구...ErrorCode에서 에러 메시지를 관리하고 있으니까 이것도 ErrorCode로 넘겨서 에러 메시지 관리하는게 좋을 것 같아요!☺️(천천히 수정하셔도됩니당👍🩷)

sooyeoneo replied:
체크해주셔서 감사합니다! 수정했습니다!!! 😁💗

jiyoon0000 commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/service/AptitudeService.java:
```diff
@@ -0,0 +1,204 @@
+		for (AptitudeTestHistory history : histories) {
+			String aiResponse = history.getAiResponse();
+			if (aiResponse != null) {
+				// 간단한 키워드 매칭 방식 (실제로는 AI 서비스에서 더 정교하게 처리)
```
지금 `AptitudeAiService`의 `analyzeResponse()`에서 조금 더 정교화된 방식으로 수정되는 걸까요?

jiyoon0000 replied:
그냥 참고만 해주세요! 이런 방식도 좋을 것 같아서 의견 남겨요!🧐

현재 로직을 보면 `contains()`로만 매칭을 하는데, 이렇게 할 경우 GPT 응답이 예상치 못하게 나왔을 경우 매칭이 안될 수도 있을 것 같아요!🥲

```
for (AptitudeTestHistory history : histories) {
    String aiResponse = history.getAiResponse();
    if (aiResponse != null) {
       
        String matchedCode = Pattern.compile("(NATURE|TECH|COMMUNITY|ART_CREATION|HISTORY_CULTURE)")
            .matcher(aiResponse)
            .results()
            .map(MatchResult::group)
            .findFirst()
            .orElse(null);

        if (matchedCode != null) {
            for (AptitudeType type : AptitudeType.values()) {
                if (matchedCode.equals(type.name())) {
                    scores.put(type, scores.get(type) + 1);
                }
            }
        }
    }
}
```

그래서 이런식으로 정규식으로 코드를 뽑아내고 enum이랑 매칭하는 방식을 사용하면 안전하지 않을까 생각해봤습니다!😎

sooyeoneo replied:
네! 현재는 AI가 자유 형식의 텍스트로 응답하고, 정규식으로 파싱하는 2단계 구조인데요,

지금보다 더 정교한 방식으로 개선한다면..? 🧐

AI 프롬프트를 수정해서 JSON 형식으로 적성 타입과 신뢰도 점수를 반환하도록 하거나
구조화된 응답 형식을 요청해서 파싱 오류를 줄이는 방향으로 개선할 수 있을 것 같습니다!!! 😆

이 부분은 현재 PR 범위가 이미 크고, 프롬프트 엔지니어링과 응답 파싱 로직 전체를 수정해야 해서, 
별도 이슈로 분리해서 진행하도록 하겠습니다! 🧑🏻‍💻💭

일단 현재는 알려주신 정규식 - enum 매칭으로 개선한 상태입니다!

```
for (AptitudeTestHistory history : histories) {
    String aiResponse = history.getAiResponse();
    if (aiResponse != null) {
       
        String matchedCode = Pattern.compile("(NATURE|TECH|COMMUNITY|ART_CREATION|HISTORY_CULTURE)")
            .matcher(aiResponse)
            .results()
            .map(MatchResult::group)
            .findFirst()
            .orElse(null);

        if (matchedCode != null) {
            for (AptitudeType type : AptitudeType.values()) {
                if (matchedCode.equals(type.name())) {
                    scores.put(type, scores.get(type) + 1);
                }
            }
        }
    }
}
```

 알찬 코멘트 정말 감사해요~😍❣️

sooyeoneo replied:
궁금한 점이 생겨서 여쭤봅니다~! 🥰

알려주신 코드로 수정했는데, 알려주신 코드가 매번 `Pattern.compile()` 호출을 하는 것 같아서 
`Pattern` 한 번만 컴파일 되는 로직으로 수정할까 하는데 어떻게 생각하시나요? 코드는 아래와 같습니다! 
(더 나은 버전으로 하고 싶어서 여쭤봅니다!)

```
	private Map<AptitudeType, Integer> analyzeHistories(List<AptitudeTestHistory> histories) {
		Map<AptitudeType, Integer> scores = new HashMap<>();
		for (AptitudeType type : AptitudeType.values()) {
			scores.put(type, 0);
		}

		// 정규식 패턴 생성 - 모든 적성 타입을 OR로 연결
		String pattern = Arrays.stream(AptitudeType.values())
			.map(type -> type.name() + "|" + type.getTitle())
			.collect(Collectors.joining("|"));
		Pattern aptitudePattern = Pattern.compile("(" + pattern + ")");

		// AI 응답에서 적성 키워드 추출 및 점수 계산
		for (AptitudeTestHistory history : histories) {
			String aiResponse = history.getAiResponse();
			if (aiResponse != null) {
				Matcher matcher = aptitudePattern.matcher(aiResponse);
				while (matcher.find()) {
					String matchedCode = matcher.group();
					// 매칭된 코드가 어떤 적성 타입인지 확인
					for (AptitudeType type : AptitudeType.values()) {
						if (matchedCode.equals(type.name()) || matchedCode.equals(type.getTitle())) {
							scores.put(type, scores.get(type) + 1);
							break;
						}
					}
				}
			}
		}

		return scores;
	}
}

j-hann commented on src/main/java/com/salayo/locallifebackend/domain/ai/aptitude/controller/AptitudeController.java:
```diff
@@ -0,0 +1,73 @@
+	private final AptitudeService aptitudeService;
+
+	@PostMapping("/test/start")
+	@PreAuthorize("hasRole('USER')")
```
저희 서비스가 로그인한 사용자만 이용 가능한 서비스라 `@PreAuthorize("hasRole('USER')")` 어노테이션은 생략하는 방향도 괜찮을 것 같아요! 

SecurityConfig 에서도 
```
.requestMatchers("/admin/**").hasRole("ADMIN")
.requestMatchers("/localcreator/**").hasRole("LOCAL_CREATOR")
```
이렇게 관리자랑 로컬 크리에이터 권한만 url로 제한하고있어서 일반 사용자가 접근 가능한 api는 어노테이션 붙이지 않아도 괜찮을 것 같다는 생각인데 의견 나눠보면 좋을 것 같아유🤗

sooyeoneo replied:
넵 `@PreAuthorize("hasRole('USER')")`은 생략하고 다시 Push 하겠습니다! 감사합니다💗😚

j-hann reviewed (approved):
짱짱👍✨

jiyoon0000 commented:
🩷👍😎너무 고생하셨어요오오오 짱이예요!!!!😎👍🩷

sooyeoneo committed:
🔥 remove : @PreAuthorize("hasRole('USER')") 제거

sooyeoneo committed:
🔧 add : NOT_FOUND_TEST_PROGRESS, TOO_MANY_REQUESTS ErrorCode에 추가

sooyeoneo committed:
🔧 add : 적성 검사 AI 응답 분석에 정규식 패턴 매칭 로직 추가

- analyzeHistories() 메서드에 Pattern/Matcher 기반 분석 로직 적용
- 정규식으로 모든 적성 타입(name + title) OR 연결 패턴 생성
- AI 응답 내 복수 적성 키워드 추출 및 점수 계산 지원
- 향후 HashMap 최적화를 위한 TODO 주석 추가

Refs: #65
